File: db/url.db.go
package db

import (
	"github.com/Tsuzat/zipit-go-fiber/config"
	"github.com/Tsuzat/zipit-go-fiber/models"
	"github.com/gofiber/fiber/v2/log"
)

func GetUrlByAlias(alias string) (*models.Url, error) {
	url := &models.Url{
		Alias: alias,
	}
	if err := config.DB.Model(url).Where("alias = ?", alias).Select(); err != nil {
		return nil, err
	}
	return url, nil
}

func GetUrlById(id int) (*models.Url, error) {
	url := &models.Url{
		Id: id,
	}
	if err := config.DB.Model(url).WherePK().Select(); err != nil {
		return nil, err
	}
	return url, nil
}

func InsertUrl(url *models.Url) error {
	_, err := config.DB.Model(url).Insert()
	if err != nil {
		return err
	}
	log.Info("Created Url with id: ", url.Id)
	return nil
}

func UpdateUrl(url *models.Url) error {
	_, err := config.DB.Model(url).Where("id = ?", url.Id).Update()
	if err != nil {
		return err
	}
	log.Info("Updated Url with id: ", url.Id)
	return nil
}

func DeleteUrl(id int) error {
	res, err := config.DB.Model(&models.Url{}).Where("id = ?", id).Delete()
	if err != nil {
		return err
	}
	if res.RowsAffected() == 1 {
		log.Info("Deleted Url with id: ", id)
	}
	return nil
}

func GetUrlsByOwnerId(owner int) ([]models.Url, error) {
	var urls []models.Url
	if err := config.DB.Model(&urls).Where("owner = ?", owner).Select(); err != nil {
		return nil, err
	}
	return urls, nil
}

func CountUrlsByOwnerId(owner int) (int, error) {
	return config.DB.Model((*models.Url)(nil)).Where("owner = ?", owner).Count()
}

File: src/database/url.db.ts
import { eq } from "drizzle-orm";
import "../models/url.model";
import db from "./db";
import { urls } from "./schemas";

export const insertUrl = async (url: Url): Promise<number> => {
  const results = await db.insert(urls).values(url).returning({ id: urls.id });
  if (results.length === 0) {
    return -1;
  }
  return results[0].id;
};

export const findUrlByAlias = async (alias: string): Promise<Url | null> => {
  const query = db.select().from(urls).where(eq(urls.alias, alias)).limit(1);
  const [url] = await query;
  return url || null;
};

export const updateUrl = async (url: Url): Promise<boolean> => {
  const result = await db
    .update(urls)
    .set(url)
    .where(eq(urls.id, url.id))
    .returning({ id: urls.id });
  if (result.length === 0) {
    return false;
  }
  return true;
};

export const deleteUrl = async (id: number): Promise<boolean> => {
  const result = await db
    .delete(urls)
    .where(eq(urls.id, id))
    .returning({ id: urls.id });
  if (result.length === 0) {
    return false;
  }
  return true;
};

export const getAllUrlsByUser = async (userId: number): Promise<Url[]> => {
  const query = db.select().from(urls).where(eq(urls.owner, userId));
  return await query;
};
